src/main/java/com/google/javascript/clutz/DeclarationGenerator.java

```diff
@@ -3097,18 +3113,21 @@ private void visitFunctionParameters(
         if (param.isVarArgs()) {
           emit("...");
         }
+
+        String pName = "";
         if (names != null && names.hasNext()) {
-          emitNoSpace(names.next());
-        } else {
-          String pName;
+          pName = names.next();
+        }
```

Comment:
Good time to add a comment here. Something like "When we can't pick a reasonable name from the AST, but the type object tells us there are parameters with known types, we generate names like p1, p2, ...".

Also add tests for destructuring functions/methods:
```
foo([a, b]) {...}
bar({x, y}) {...}
```